### PR TITLE
'Code Together' Create a separate requirements file for the CI issue #1292

### PR DESCRIPTION
**Pull request by pratjz**
------------------------------------------------------------------------------------------------------------

**Context:** 
The requirements.txt file has packages unpinned. This is contrary to the convention of pinning packages in the requirements.txt file. The reason for this is that the CI installs the environment from this file, and it helps catching bugs/issues when there are dependency upgrades. 

**Description of the Change:**
To have a separate requirements.txt (pinned) and requirements-ci.txt (unpinned). This latter would be used by the CI.

- Create a requirements-ci.txt file in the top-level pennylane directory where the requirements.txt exists. This file should be a copy of requirements.txt, but should not pin the version number of the packages contained within.

- Pin the packages that are unpinned in the requirements.txt file to the latest released version of the given packages. E.g., numpy should become numpy==1.21

- Change the usage of requirements.txt in files that are located in the .github/workflows directory to use the new requirements-ci.txt file.

**Benefits:**
The requirements.txt file has packages unpinned. This is contrary to the convention of pinning packages in the requirements.txt file. The reason for this is that the CI installs the environment from this file, and it helps catching bugs/issues when there are dependency upgrades.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by rmoyard** [comment]
Hi @pratjz, thank you very much for the quick PR and participating to the event! We will review it soon :)

File: requirements-ci.txt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+numpy
+scipy
+cvxpy
+cvxopt
+networkx
+tensornetwork
```
Comment (rmoyard) [inline requirements-ci.txt:6]:
```suggestion
tensornetwork==0.3
```

File: requirements-ci.txt [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+autograd
+toml
+appdirs
+semantic_version
```
Comment (rmoyard) [inline requirements-ci.txt:10]:
```suggestion
semantic_version==2.6
```

File: requirements.txt [hunk 3]
```diff
@@ -1,13 +1,13 @@
...
-semantic_version==2.6
-dask[delayed]
-autoray
-matplotlib
+numpy=1.21
+scipy=1.7
+cvxpy=1.1
+cvxopt=1.2
+networkx=2.6
+tensornetwork=0.4
```
Comment (rmoyard) [inline requirements.txt:6]:
```suggestion
tensornetwork==0.3
```

File: requirements.txt [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
+autograd=1.0
+toml=1.0
+appdirs=1.4
+semantic_version=2.8
```
Comment (rmoyard) [inline requirements.txt:10]:
```suggestion
semantic_version=2.6
```

File: requirements.txt [hunk 5]
```diff
@@ -1,13 +1,13 @@
...
-networkx
-tensornetwork==0.3
-autograd
-toml
-appdirs
-semantic_version==2.6
-dask[delayed]
-autoray
-matplotlib
+numpy=1.21
```
Comment (rmoyard) [inline requirements.txt:1]:
```suggestion
numpy==1.21
```

File: requirements.txt [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
+scipy=1.7
```
Comment (rmoyard) [inline requirements.txt:2]:
```suggestion
scipy==1.7
```

File: requirements.txt [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
+cvxpy=1.1
```
Comment (rmoyard) [inline requirements.txt:3]:
```suggestion
cvxpy==1.1
```

File: requirements.txt [hunk 8]
```diff
@@ -1,13 +1,13 @@
...
+cvxopt=1.2
```
Comment (rmoyard) [inline requirements.txt:4]:
```suggestion
cvxopt==1.2
```

File: requirements.txt [hunk 9]
```diff
@@ -1,13 +1,13 @@
...
+networkx=2.6
```
Comment (rmoyard) [inline requirements.txt:5]:
```suggestion
networkx==2.6
```

File: requirements.txt [hunk 10]
```diff
@@ -1,13 +1,13 @@
...
+tensornetwork=0.4
+autograd=1.0
```
Comment (rmoyard) [inline requirements.txt:7]:
```suggestion
autograd==1.0
```

File: requirements.txt [hunk 11]
```diff
@@ -1,13 +1,13 @@
...
+toml=1.0
```
Comment (rmoyard) [inline requirements.txt:8]:
```suggestion
toml==1.0
```

File: requirements.txt [hunk 12]
```diff
@@ -1,13 +1,13 @@
...
+appdirs=1.4
```
Comment (rmoyard) [inline requirements.txt:9]:
```suggestion
appdirs==1.4
```

File: requirements.txt [hunk 13]
```diff
@@ -1,13 +1,13 @@
...
+semantic_version=2.8
+dask[delayed]=2021.08
```
Comment (rmoyard) [inline requirements.txt:11]:
```suggestion
dask[delayed]==2021.08
```

File: requirements.txt [hunk 14]
```diff
@@ -1,13 +1,13 @@
...
+autoray=0.2
```
Comment (rmoyard) [inline requirements.txt:12]:
```suggestion
autoray==0.2
```

File: requirements.txt [hunk 15]
```diff
@@ -1,13 +1,13 @@
...
+matplotlib=3.4
```
Comment (rmoyard) [inline requirements.txt:13]:
```suggestion
matplotlib==3.4
```

**Review by rmoyard — COMMENTED** [review]
Thanks for the PR @pratjz! I've left some suggestions, two packages should stayed pinned anyway (tensornetwork and semantic_version), it should solved the non passing tests.

**Comment by pratjz** [comment]
> Thanks for the PR @pratjz! I've left some suggestions, two packages should stayed pinned anyway (tensornetwork and semantic_version), it should solved the non passing tests. Pls also add your name in the list of contributors in `CHANGELOG.md` .

Thank You rmoyard for the quick feedback

File: requirements.txt [hunk 16]
```diff
@@ -1,13 +1,13 @@
-numpy
-scipy
-cvxpy
-cvxopt
-networkx
+numpy==1.21
+scipy==1.7
+cvxpy==1.1
+cvxopt==1.2
+networkx==2.6
```
Comment (rmoyard) [inline requirements.txt:5]:
```suggestion
numpy==1.21.2
scipy==1.7.1
cvxpy==1.1.15
cvxopt==1.2.6
networkx==2.6.2
```

File: requirements.txt [hunk 17]
```diff
@@ -1,13 +1,13 @@
...
 tensornetwork==0.3
-autograd
-toml
-appdirs
+autograd==1.0
+toml==1.0
+appdirs==1.4
```
Comment (rmoyard) [inline requirements.txt:9]:
```suggestion
autograd==1.3
toml==0.10.2
appdirs==1.4.4
```

File: requirements.txt [hunk 18]
```diff
@@ -1,13 +1,13 @@
...
 semantic_version==2.6
-dask[delayed]
-autoray
-matplotlib
+dask[delayed]==2021.08
+autoray==0.2
+matplotlib==3.4
```
Comment (rmoyard) [inline requirements.txt:13]:
```suggestion
autoray==0.2.5
matplotlib==3.4.3
```

**Review by rmoyard — COMMENTED** [review]
Hi, @pratjz thanks for the update! I have updated some of the package versions 💯

File: .github/CHANGELOG.md [hunk 19]
```diff
@@ -121,6 +121,11 @@
   an owner of the pruned tensor and its constituent observables, but leaves the
   original tensor in the queue without an owner.
 
+* Create a separate requirements file for the CI issue , to have a separate requirements.txt (pinned)
+and requirements-ci.txt (unpinned). This latter would be used by the CI..
```
Comment (antalszava) [inline .github/CHANGELOG.md:125]:
```suggestion
and requirements-ci.txt (unpinned). This latter would be used by the CI.
```

File: .github/CHANGELOG.md [hunk 20]
```diff
@@ -131,7 +136,7 @@
 
 This release contains contributions from (in alphabetical order):
 
-Josh Izaac, Johannes Jakob Meyer, Maria Schuld.
+Josh Izaac, Johannes Jakob Meyer, Maria Schuld, Prateek Jain,
```
Comment (antalszava) [inline .github/CHANGELOG.md:139]:
```suggestion
Josh Izaac, Prateek Jain, Johannes Jakob Meyer, Maria Schuld.
```

**Comment by rmoyard** [comment]
Hey @pratjz, I think you are mixing different PRs with your last pushes ?

**Comment by pratjz** [comment]
> Hey @pratjz, I think you are mixing different PRs with your last pushes ?

I was trying to create another PR for another issue, but had not created it yet, is it showing those commits here?

**Comment by pratjz** [comment]
> > Hey @pratjz, I think you are mixing different PRs with your last pushes ?
> 
> I was trying to create another PR for another issue, but had not created it yet, is it showing those commits here?

Ya I can see it showing those commits here too, where did I mess up

**Comment by pratjz** [comment]
Hey @rmoyard I have reverted the commits from other PR now , Kindly review

**Comment by rmoyard** [comment]
Thanks @pratjz! I am looking forward for your new PR

**Comment by pratjz** [comment]
> Thanks @pratjz! I am looking forward for your new PR

Hey @rmoyard one quick doubt probably stupid one, now I made another branch for another issue and trying to make PR for that, but it is showing file changes from this PR into that also, is it like I can't open another PR until this one is merged ?

**Comment by rmoyard** [comment]
@pratjz You should create a new branch from the master one not from this one. Work on the new branch for the new project.

**Comment by pratjz** [comment]
> @pratjz You should create a new branch from the master one not from this one. Work on the new branch for the new project.

Got it Thank You

**Review by rmoyard — APPROVED** [review]
Thanks for your good work :) I approve the pull request 💯 We will contact you in private shortly.

**Comment by pratjz** [comment]
> Thanks for your good work :) I approve the pull request 💯 We will contact you in private shortly.

Thank You @rmoyard

**Pull request merged**